### PR TITLE
Adding message builder for expected event strings 

### PR DESCRIPTION
**Pull request by n3wscott**
Related to #1393

### CODE REVIEW

**Comment by n3wscott** [comment]
Also related to #1409

File: pkg/controller/controller_binding_test.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func TestReconcileServiceBindingNonExistingServiceInstance(t *testing.T) {
 	// There should only be one action that says it failed because no such instance exists.
 	updateAction := actions[0].(clientgotesting.UpdateAction)
 	if e, a := "update", updateAction.GetVerb(); e != a {
-		t.Fatalf("Unexpected verb on actions[0]; expected %v, got %v", e, a)
+		t.Fatalf("Unexpected verb on actions[0]; %s", expectedGot(e, a))
```
Comment (pmorie) [inline pkg/controller/controller_binding_test.go]: This seems unrelated to the rest of this PR - am I missing something?

File: pkg/controller/controller_binding_test.go [hunk 2]
```diff
@@ -82,8 +82,9 @@ func TestReconcileServiceBindingNonExistingServiceInstance(t *testing.T) {
 	events := getRecordedEvents(testController)
 	assertNumEvents(t, events, 1)
 
-	expectedEvent := corev1.EventTypeWarning + " " + errorNonexistentServiceInstanceReason + " " + "References a non-existent ServiceInstance \"/nothere\""
-	if err := checkEvents(events, []string{expectedEvent}); err != nil {
+	expectedEvent := new(MessageBuilder).warning().reason(errorNonexistentServiceInstanceReason)
+	expectedEvent.msgf("References a non-existent ServiceInstance %q", "/"+testNonExistentClusterServiceClassName)
+	if err := checkEvents(events, expectedEvent.stringArr()); err != nil {
```
Comment (pmorie) [inline pkg/controller/controller_binding_test.go]: why `stringArr` instead of just `String()`?

File: pkg/controller/controller_message_builder_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+// msgf appends a message after formatting to the message builder
+func (mb *MessageBuilder) msgf(format string, a ...interface{}) *MessageBuilder {
+	msg := fmt.Sprintf(format, a...)
+	return mb.msg(msg)
+}
+
+// msgCreateServiceBindingError adds a message in the following form:
+// "Error creating ServiceBinding for ServiceInstance %q of ClusterServiceClass (K8S: %q ExternalName: %q) at ClusterServiceBroker %q:"
+func (mb *MessageBuilder) msgCreateServiceBindingError(serviceInstance, serviceClassK8S, serviceClassExternalName, broker string) *MessageBuilder {
```
Comment (pmorie) [inline pkg/controller/controller_message_builder_test.go]: What's the decision plane for when to create one of these special-purpose methods?

Comment (pmorie) [reply pkg/controller/controller_message_builder_test.go]: My gut is that the special-purpose methods might turn into more trouble than they're worth.

File: pkg/controller/controller_binding_test.go (hunk 2 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: it needs to be a []string

File: pkg/controller/controller_binding_test.go (hunk 1 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: I fixed this because I broke something in that test but the expected/got format was not updated and it was hard to debug so I fixed it as I went, but just this one. I can pull it from the PR if needed

File: pkg/controller/controller_message_builder_test.go (hunk 3 above)
Comment (n3wscott) [reply pkg/controller/controller_message_builder_test.go]: It was used more that a few times and it was tedious and error prone to bring along the string each time. I was finding that minor copy/pasta errors were causing tests to break and making a msg* function really helped out.

I would not do this for production type code, but for tests I think it feels appropriate.

**Comment by n3wscott** [comment]
> WDYT

I think we can remove the helpers if needed, the string is quite long with all the parameters so that was what this was solving. Making it easier to read. The wrapped messages are in the form "Blah blah yada:" and are always appended with an error message after making it "Blah blah yada: <error>" And in the case of the unit tests <error> tends to be the interesting bit, not the blah blah part. Hence the convenience method.

**Comment by pmorie** [comment]
> I think we can remove the helpers if needed, the string is quite long with all the parameters so that was what this was solving. Making it easier to read. The wrapped messages are in the form "Blah blah yada:" and are always appended with an error message after making it "Blah blah yada: " And in the case of the unit tests tends to be the interesting bit, not the blah blah part. Hence the convenience method.

I can live with that

**Comment by n3wscott** [comment]
Updated and simplified the builder.

File: pkg/controller/controller_binding_test.go [hunk 4]
```diff
@@ -241,8 +242,12 @@ func TestReconcileServiceBindingNonExistingClusterServiceClass(t *testing.T) {
 	events := getRecordedEvents(testController)
 	assertNumEvents(t, events, 1)
 
-	expectedEvent := corev1.EventTypeWarning + " " + errorNonexistentClusterServiceClassMessage + " " + "References a non-existent ClusterServiceClass (K8S: \"nosuchclassid\" ExternalName: \"" + testNonExistentClusterServiceClassName + "\")"
-	if err := checkEvents(events, []string{expectedEvent}); err != nil {
+	expectedEvent := new(MessageBuilder).warning().reason(errorNonexistentClusterServiceClassMessage)
```
Comment (pmorie) [inline pkg/controller/controller_binding_test.go]: I am liking where this is heading - one thought -

How about instead of:

```go
expectedEvent := new(MessageBuilder).warning().reason(errorNonexistentClusterServiceClassMessage)
```

we have something like:

```go
expectedEvent := warningEvent().reason(errorNonexistentClusterServiceClassMessage)
```

Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Like a function in the controller test that wraps and returns the `new(MessageBuilder).warning()`

how about `warningEventBuilder().yada...` to make it clear it is still this builder object?

Comment (pmorie) [reply pkg/controller/controller_binding_test.go]: I could live with `warningEventBuilder()` over `new(MessageBuilder).warning()`

**Comment by n3wscott** [comment]
Tests are now much more compact, check it out! 😎
